File: btils/bfmt.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <filesystem>
#include <stdexcept>

namespace fs = std::filesystem;

std::string formatLine(const std::string& line) {
    if (line.find(';') != std::string::npos) return "";
    if (line.back() == ':' || line.front() == '.') return line;
    return "    " + line;
}

std::vector<std::string> readFileSafe(const std::string& filePath) {
    std::ifstream file(filePath);
    if (!file) throw std::runtime_error("Error reading file");
    
    std::vector<std::string> lines;
    std::string line;
    while (std::getline(file, line)) {
        lines.push_back(line);
    }
    return lines;
}

void writeFile(const std::string& filePath, const std::vector<std::string>& lines) {
    std::ofstream file(filePath);
    for (const auto& line : lines) {
        file << line << "\n";
    }
}

void processInputFile(const std::string& inputFilePath) {
    if (!fs::exists(inputFilePath)) return;

    try {
        auto fileContents = readFileSafe(inputFilePath);
        std::vector<std::string> formattedLines;
        for (const auto& line : fileContents) {
            formattedLines.push_back(formatLine(line));
        }

        std::string tempFilePath = inputFilePath + ".tmp";
        writeFile(tempFilePath, formattedLines);

        fs::rename(tempFilePath, inputFilePath);
    } catch (const std::exception& e) {
        std::cerr << e.what() << std::endl;
        if (fs::exists(inputFilePath + ".tmp")) {
            fs::remove(inputFilePath + ".tmp");
        }
    }
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cout << "Usage: " << argv[0] << " <input>" << std::endl;
        return 1;
    }

    for (int i = 1; i < argc; ++i) {
        processInputFile(argv[i]);
    }

    return 0;
}

File: btils/bfmt.hs
import System.IO
import System.Environment (getArgs, getProgName)
import System.Directory (renameFile, doesFileExist, removeFile)
import Data.List (isPrefixOf, isSuffixOf, isInfixOf)
import System.FilePath (takeBaseName, takeDirectory)
import Control.Exception (catch, SomeException)

formatLine :: String -> String
formatLine line
  | ";" `isInfixOf` line = ""
  | isSuffixOf ":" line || isPrefixOf "." line = line
  | otherwise = "    " ++ line

main :: IO ()
main = do
  commandLineArgs <- getArgs
  if null commandLineArgs
    then do
      progName <- getProgName
      putStrLn $ "Usage: " ++ progName ++ " <input>"
      return ()
    else mapM_ processInputFile commandLineArgs

processInputFile :: FilePath -> IO ()
processInputFile inputFilePath = do
  fileExists <- doesFileExist inputFilePath
  if not fileExists
    then return ()
    else do
      contents <- readFileSafe inputFilePath
      case contents of
        Left _ -> return ()
        Right fileContents -> do
          let formattedLines = map formatLine (lines fileContents)
          let tempFilePath = takeDirectory inputFilePath ++ "/" ++ takeBaseName inputFilePath ++ ".tmp"
          writeFile tempFilePath (unlines formattedLines)
          renameResult <- tryRename tempFilePath inputFilePath
          case renameResult of
            Left _ -> removeFileSafe tempFilePath
            Right _ -> return ()

readFileSafe :: FilePath -> IO (Either String String)
readFileSafe filePath = catch (Right <$> readFile filePath) handleError
  where
    handleError :: SomeException -> IO (Either String String)
    handleError _ = return (Left "Error reading file")

tryRename :: FilePath -> FilePath -> IO (Either String ())
tryRename tempFilePath originalFilePath = catch (Right <$> renameFile tempFilePath originalFilePath) handleError
  where
    handleError :: SomeException -> IO (Either String ())
    handleError _ = return (Left "Error renaming file")

removeFileSafe :: FilePath -> IO ()
removeFileSafe filePath = catch (removeFile filePath) handleError
  where
    handleError :: SomeException -> IO ()
    handleError _ = return ()
